### PR TITLE
API calls fixes

### PR DESCRIPTION
**Pull request by iulianpascalau**
- implemented a new function in the chain handlers implementations that can provide the last committed roothash so the API calls can function as expected
- cleanup in the chain handler stubs & mocks

testing scenario: 
1. on a regular internal testnet with high load and generated scheduled miniblocks, repeatedly call the route `/address/:address/esdt` for an account that contains ESDTs. This will seldom work, most of the time returning `get esdt tokens for account error: key not found` error.
2. repeat the process on this branch. The calls should work each time

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -9,7 +9,7 @@ require (
 	github.com/ElrondNetwork/concurrent-map v0.1.3
 	github.com/ElrondNetwork/covalent-indexer-go v1.0.6
 	github.com/ElrondNetwork/elastic-indexer-go v1.1.33
-	github.com/ElrondNetwork/elrond-go-core v1.1.9
+	github.com/ElrondNetwork/elrond-go-core v1.1.11-0.20220202190944-14577a435830
```
Comment (sasurobert) [inline go.mod:12]: proper release needed

File: process/smartContract/hooks/blockChainHook.go [hunk 2]
```diff
@@ -299,6 +299,7 @@ func (bh *BlockChainHookImpl) LastEpoch() uint32 {
 
 // GetStateRootHash returns the state root hash from the last committed block
 func (bh *BlockChainHookImpl) GetStateRootHash() []byte {
+	// TODO decide here, in this PR if we should use the block's root hash or the committed root hash from the chain handler
 	if !check.IfNil(bh.blockChain.GetCurrentBlockHeader()) {
```
Comment (sasurobert) [inline process/smartContract/hooks/blockChainHook.go:303]: it is better to choose the commited hash.

Comment (iulianpascalau) [reply process/smartContract/hooks/blockChainHook.go]: done

File: go.mod (hunk 1 above)
Comment (iulianpascalau) [reply go.mod]: done

File: process/block/shardblock.go [hunk 3]
```diff
@@ -932,7 +932,12 @@ func (sp *shardProcessor) CommitBlock(
 
 	lastBlockHeader := sp.blockChain.GetCurrentBlockHeader()
 
-	err = sp.blockChain.SetCurrentBlockHeader(header)
+	commitedRoothash, err := sp.accountsDB[state.UserAccountsState].RootHash()
```
Comment (AdoAdoAdo) [inline process/block/shardblock.go:935]: :+1:

File: process/factory/shard/vmContainerFactory_test.go [hunk 4]
```diff
@@ -1,3 +1,4 @@
+//go:build !race
```
Comment (AdoAdoAdo) [inline process/factory/shard/vmContainerFactory_test.go:1]: why do we have two go compiler directives here?

go:build !race and +build !race ?

**Review by sasurobert — DISMISSED** [review]

File: process/block/metablock.go [hunk 5]
```diff
@@ -1247,7 +1247,7 @@ func (mp *metaProcessor) CommitBlock(
 	lastMetaBlockHash := mp.blockChain.GetCurrentBlockHeaderHash()
 	mp.updateState(lastMetaBlock, lastMetaBlockHash)
 
-	err = mp.blockChain.SetCurrentBlockHeader(header)
+	err = mp.blockChain.SetCurrentBlockHeaderAndRootHash(header, header.RootHash)
```
Comment (AdoAdoAdo) [inline process/block/metablock.go:1250]: maybe get here as well the committed roothash? 
to cover also the case where we add scheduled System SC calls on metachain

**Review by AdoAdoAdo — APPROVED** [review]

File: process/factory/shard/vmContainerFactory_test.go (hunk 4 above)
Comment (iulianpascalau) [reply process/factory/shard/vmContainerFactory_test.go:1]: removed the compiler directives & also the TODO. Arwen is fixed

File: process/block/metablock.go (hunk 5 above)
Comment (iulianpascalau) [reply process/block/metablock.go:1250]: done

**Review by AdoAdoAdo — APPROVED** [review]

File: process/smartContract/hooks/blockChainHook.go [hunk 6]
```diff
@@ -261,8 +265,12 @@ func (bh *BlockChainHookImpl) GetBlockhash(nonce uint64) ([]byte, error) {
 	return hash, nil
 }
 
-// LastNonce returns the nonce from from the last committed block
+// LastNonce returns the nonce from the last committed block
 func (bh *BlockChainHookImpl) LastNonce() uint64 {
+	if bh.flagDisableBlockchainOperations.IsSet() {
```
Comment (sasurobert) [inline process/smartContract/hooks/blockChainHook.go:270]: Last"*" operations are needed and they are good

File: process/smartContract/hooks/blockChainHook.go [hunk 7]
```diff
@@ -303,10 +327,16 @@ func (bh *BlockChainHookImpl) LastEpoch() uint32 {
 
 // GetStateRootHash returns the state root hash from the last committed block
 func (bh *BlockChainHookImpl) GetStateRootHash() []byte {
-	if !check.IfNil(bh.blockChain.GetCurrentBlockHeader()) {
-		return bh.blockChain.GetCurrentBlockHeader().GetRootHash()
+	if bh.flagDisableBlockchainOperations.IsSet() {
```
Comment (sasurobert) [inline process/smartContract/hooks/blockChainHook.go:330]: please leave this.

File: process/smartContract/hooks/blockChainHook.go [hunk 8]
```diff
@@ -287,14 +303,22 @@ func (bh *BlockChainHookImpl) LastTimeStamp() uint64 {
...
 	if !check.IfNil(bh.blockChain.GetCurrentBlockHeader()) {
 		return bh.blockChain.GetCurrentBlockHeader().GetRandSeed()
 	}
-	return []byte{}
+	return make([]byte, 0)
 }
 
 // LastEpoch returns the epoch from the last committed block
 func (bh *BlockChainHookImpl) LastEpoch() uint32 {
+	if bh.flagDisableBlockchainOperations.IsSet() {
```
Comment (sasurobert) [inline process/smartContract/hooks/blockChainHook.go:318]: the only one which needs disable is the GetBlockhash(nonce uint64)

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: process/smartContract/hooks/blockChainHook.go (hunk 8 above)
Comment (iulianpascalau) [reply process/smartContract/hooks/blockChainHook.go]: done

File: process/smartContract/hooks/blockChainHook.go (hunk 7 above)
Comment (iulianpascalau) [reply process/smartContract/hooks/blockChainHook.go]: done

File: process/smartContract/hooks/blockChainHook.go (hunk 6 above)
Comment (iulianpascalau) [reply process/smartContract/hooks/blockChainHook.go]: done

**Review by AdoAdoAdo — DISMISSED** [review]

File: process/smartContract/hooks/blockChainHook.go [hunk 9]
```diff
@@ -683,6 +696,9 @@ func (bh *BlockChainHookImpl) EpochConfirmed(epoch uint32, _ uint64) {
 
 	bh.flagOptimizeNFTStore.SetValue(epoch >= bh.optimizeNFTStoreEnableEpoch)
 	log.Debug("blockchainHookImpl optimize nft metadata store", "enabled", bh.flagOptimizeNFTStore.IsSet())
+
+	bh.flagDoNotReturnOldBlock.SetValue(epoch >= bh.doNotReturnOldBlockEnableEpoch)
+	log.Debug("blockchainHook: do not return old block", "enabled", bh.flagDoNotReturnOldBlock.IsSet())
```
Comment (SebastianMarian) [inline process/smartContract/hooks/blockChainHook.go:701]: blockchainHookImpl

File: process/sync/baseSync.go [hunk 10]
```diff
@@ -924,12 +931,13 @@ func (boot *baseBootstrap) rollBackToNonceForced() {
 func (boot *baseBootstrap) restoreState(
 	currHeaderHash []byte,
 	currHeader data.HeaderHandler,
+	currRootHash []byte,
 ) {
 	log.Debug("revert state to header",
 		"nonce", currHeader.GetNonce(),
 		"hash", currHeaderHash)
```
Comment (SebastianMarian) [inline process/sync/baseSync.go:938]: Add also currRootHash in log

**Review by SebastianMarian — DISMISSED** [review]

File: process/smartContract/hooks/blockChainHook.go (hunk 9 above)
Comment (iulianpascalau) [reply process/smartContract/hooks/blockChainHook.go:701]: done

File: process/sync/baseSync.go (hunk 10 above)
Comment (iulianpascalau) [reply process/sync/baseSync.go:938]: done

**Review by SebastianMarian — APPROVED** [review]

**Review by AdoAdoAdo — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
System test passed.

**Pull request merged**